File: matlab/+spine/+systems/system_spin.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef system_spin < handle

        properties
            
            % Number of dots
            dots;
            
            % Hamiltonian Properties
            larmor_frequency;
            rabi_frequency;
            charging_energy;
            singlet_triplet_energy;

            % Hamiltonian Control
            microwave_control;
            detuning_control;
            tunnel_control;
            
            % Plot handles
            p;
            
        end
        
        properties (Constant)
            
            % Supported spin states:
            STATE_N = 0              % 0 electrons
            STATE_0 = 1              % 1 electron in ground state
            STATE_1 = 2              % 1 electron in excited state
            STATE_S = 3              % 2 electrons in singlet configuration (or in 00 spin state, depending on the Hamiltonian)
            STATE_T0 = 4             % 2 electrons in triplet (0) configuration (or in 01 spin state, depending on the Hamiltonian)
            STATE_TP = 5             % 2 electrons in triplet (+) configuration (or in 10 spin state, depending on the Hamiltonian)
            STATE_TM = 6             % 2 electrons in triplet (-) configuration (or in 11 spin state, depending on the Hamiltonian)
            
        end
        
        methods (Abstract)
            getIndex(obj, states)
            getIndexMeasurement(obj, dot, state)
            getDimension(obj)
        end
        
        methods
           
            function obj = system_spin(dots)
                obj.dots = dots;
                obj.larmor_frequency = zeros(1, dots);
                obj.rabi_frequency = zeros(1, dots);
                obj.charging_energy = zeros(1, dots);
                obj.singlet_triplet_energy = zeros(1, dots);
                obj.microwave_control = zeros(1, dots);
                obj.detuning_control = zeros(1, dots);
                obj.tunnel_control = zeros(dots, dots);
                obj.p = gobjects(1, 2 * dots);
            end
            
            function state = initialize(obj)
                state = zeros(obj.getDimension(), 1);
                states = obj.STATE_0 * ones(1, obj.dots);
                state(obj.getIndex(states)) = 1;
            end
            
            function [Px, Py, Pz] = measure(obj, state, varargin)
                
                % Measure in rotating frame at time t if requested
                if (nargin == 3)
                    t = varargin{1};
                end

                % For every dot
                Px = zeros(1, obj.dots);
                Py = zeros(1, obj.dots);
                Pz = zeros(1, obj.dots);
                for dot=1:obj.dots
                    if (nargin == 3)
                        rot_x = cos(obj.larmor_frequency(dot) * t) + 1i * sin(obj.larmor_frequency(dot) * t);
                        rot_y = cos(obj.larmor_frequency(dot) * t + pi/2) + 1i * sin(obj.larmor_frequency(dot) * t + pi/2);
                    else
                        rot_x = 1;
                        rot_y = 1i;
                    end
                    
                    % Sum over all indices where this dot in the requested state
                    Px(dot) = sum(abs((state(obj.getIndexMeasurement(dot, obj.STATE_0)) + rot_x * state(obj.getIndexMeasurement(dot, obj.STATE_1)))/sqrt(2)).^2);
                    Py(dot) = sum(abs((state(obj.getIndexMeasurement(dot, obj.STATE_0)) + rot_y * state(obj.getIndexMeasurement(dot, obj.STATE_1)))/sqrt(2)).^2);
                    Pz(dot) = sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_0))).^2);
                end
            end
            
            function [Pn, Ps, Pt] = measureST(obj, state)
                
                % For every dot
                Pn = zeros(1, obj.dots);
                Ps = zeros(1, obj.dots);
                Pt = zeros(1, obj.dots);
                for dot=1:obj.dots
                    
                    % Sum over all indices where this dot in the requested state
                    Pn(dot) = sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_N))).^2);
                    Ps(dot) = sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_S))).^2);
                    Pt(dot) = sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_T0))).^2) + ...
                              sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_TP))).^2) + ...
                              sum(abs(state(obj.getIndexMeasurement(dot, obj.STATE_TM))).^2);
                end
            end
            
            function plot(obj, state_or_U, t, varargin)
                
                % Plot style lab
                % 0: No lab frame
                % 1: Arrow lab frame
                % 2: Trace lab frame
                plot_style_lab = 0;
                if (nargin == 4)
                    plot_style_lab = varargin{1};
                end
                
                % Plot singlet-triplet
                % 0: No additional plot
                % 1: S occupancy
                % 2: 1+S+T-N occupancy (expected number of electrons)
                plot_st = 0;
                if (nargin == 5)
                    plot_st = varargin{2};
                end
                
                % Init, rotate, measure
                if (size(state_or_U, 2) == 1)
                    state = state_or_U;
                else
                    state = state_or_U * obj.initialize();
                end
                if (plot_style_lab ~= 0)
                    [Px, Py, Pz] = obj.measure(state);
                end
                [Pxr, Pyr, Pzr] = obj.measure(state, t);
                if (plot_st ~= 0)
                    [Pn, Ps, Pt] = obj.measureST(state);
                end
                
                % Plot
                if (isempty(fieldnames(get(obj.p))))
                    figure();
                    for dot=1:obj.dots
                        subplot(floor(sqrt(obj.dots)), ceil(sqrt(obj.dots)), dot);
                        spine.plotBlochSphere();
                        if (plot_style_lab == 1)
                            obj.p(2*(dot-1)+1) = plot3([0, 2*Px(dot)-1], [0, 2*Py(dot)-1], [0, 2*Pz(dot)-1]);
                        elseif (plot_style_lab == 2)
                            obj.p(2*(dot-1)+1) = plot3(2*Px(dot)-1, 2*Py(dot)-1, 2*Pz(dot)-1);
                        else
                            obj.p(2*(dot-1)+1) = plot3(0, 0, 0);
                        end
                        obj.p(2*(dot-1)+2) = plot3(2*Pxr(dot)-1, 2*Pyr(dot)-1, 2*Pzr(dot)-1);
                    end
                    if (plot_st ~= 0)
                        figure();
                        for dot=1:obj.dots
                            subplot(floor(sqrt(obj.dots)), ceil(sqrt(obj.dots)), dot);
                            if (plot_st == 1)
                                obj.p(2*obj.dots+(dot-1)+1) = plot(t, Ps(dot));
                                ylim([0, 1]);
                            elseif (plot_st == 2)
                                obj.p(2*obj.dots+(dot-1)+1) = plot(t, 1-Pn(dot)+Ps(dot)+Pt(dot));
                                ylim([0, 2]);
                            end
                        end
                    end
                else
                    for dot=1:obj.dots
                        if (plot_style_lab == 1)
                            obj.p(2*(dot-1)+1).XData = [0, 2*Px(dot)-1];
                            obj.p(2*(dot-1)+1).YData = [0, 2*Py(dot)-1];
                            obj.p(2*(dot-1)+1).ZData = [0, 2*Pz(dot)-1];
                        elseif (plot_style_lab == 2)
                            obj.p(2*(dot-1)+1).XData = [obj.p(2*(dot-1)+1).XData, 2*Px(dot)-1];
                            obj.p(2*(dot-1)+1).YData = [obj.p(2*(dot-1)+1).YData, 2*Py(dot)-1];
                            obj.p(2*(dot-1)+1).ZData = [obj.p(2*(dot-1)+1).ZData, 2*Pz(dot)-1];
                        end
                        obj.p(2*(dot-1)+2).XData = [obj.p(2*(dot-1)+2).XData, 2*Pxr(dot)-1];
                        obj.p(2*(dot-1)+2).YData = [obj.p(2*(dot-1)+2).YData, 2*Pyr(dot)-1];
                        obj.p(2*(dot-1)+2).ZData = [obj.p(2*(dot-1)+2).ZData, 2*Pzr(dot)-1];
                        if (plot_st == 1)
                            obj.p(2*obj.dots+(dot-1)+1).XData = [obj.p(2*obj.dots+(dot-1)+1).XData, t];
                            obj.p(2*obj.dots+(dot-1)+1).YData = [obj.p(2*obj.dots+(dot-1)+1).YData, Ps(dot)];
                        elseif (plot_st == 2)
                            obj.p(2*obj.dots+(dot-1)+1).XData = [obj.p(2*obj.dots+(dot-1)+1).XData, t];
                            obj.p(2*obj.dots+(dot-1)+1).YData = [obj.p(2*obj.dots+(dot-1)+1).YData, 1-Pn(dot)+Ps(dot)+Pt(dot)];
                        end
                    end
                end
                
            end
            
            % Hamiltonian Property Getters/Setters
            % set(value)        : set to all dots (could be 1 dot)
            % set(dot, value)   : set to single dot
            % get()             : get value (for 1 dot, so dot 1)
            % get(dot)          : get value of single dot
            function setLarmorFrequency(obj, varargin)
                if (nargin == 2)
                    obj.larmor_frequency = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.larmor_frequency(dot) = varargin{2};
                end
            end

            function larmor_frequency = getLarmorFrequency(obj, varargin)
                if (nargin == 1)
                    larmor_frequency = obj.larmor_frequency(1);
                else
                    dot = varargin{1};
                    larmor_frequency = obj.larmor_frequency(dot);
                end
            end

            function setRabiFrequency(obj, varargin)
                if (nargin == 2)
                    obj.rabi_frequency = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.rabi_frequency(dot) = varargin{2};
                end
            end

            function rabi_frequency = getRabiFrequency(obj, varargin)
                if (nargin == 1)
                    rabi_frequency = obj.rabi_frequency(1);
                else
                    dot = varargin{1};
                    rabi_frequency = obj.rabi_frequency(dot);
                end
            end

            function setChargingEnergy(obj, varargin)
                if (nargin == 2)
                    obj.charging_energy = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.charging_energy(dot) = varargin{2};
                end
            end

            function charging_energy = getChargingEnergy(obj, varargin)
                if (nargin == 1)
                    charging_energy = obj.charging_energy(1);
                else
                    dot = varargin{1};
                    charging_energy = obj.charging_energy(dot);
                end
            end
            
            function setSingletTripletEnergy(obj, varargin)
                if (nargin == 2)
                    obj.singlet_triplet_energy = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.singlet_triplet_energy(dot) = varargin{2};
                end
            end

            function singlet_triplet_energy = getSingletTripletEnergy(obj, varargin)
                if (nargin == 1)
                    singlet_triplet_energy = obj.singlet_triplet_energy(1);
                else
                    dot = varargin{1};
                    singlet_triplet_energy = obj.singlet_triplet_energy(dot);
                end
            end
            
            % Hamiltonian Control Setters
            function setMicrowaveControl(obj, varargin)
                if (nargin == 2)
                    obj.microwave_control = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.microwave_control(dot) = varargin{2};
                end
            end

            function setDetuningControl(obj, varargin)
                if (nargin == 2)
                    obj.detuning_control = varargin{1} * ones(1, obj.dots);
                else
                    dot = varargin{1};
                    obj.detuning_control(dot) = varargin{2};
                end
            end

            function setTunnelControl(obj, varargin)
                if (nargin == 2)
                    obj.tunnel_control = varargin{1} * ones(obj.dots, obj.dots);
                else
                    dota = varargin{1};
                    dotb = varargin{2};
                    obj.tunnel_control(dota, dotb) = varargin{3};
                    obj.tunnel_control(dotb, dota) = varargin{3};
                end
            end
            
        end
    
end

File: cpp/spine/systems/system_spin.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

#include "system_spin.h"

using namespace spine;

spine::systems::system_spin::system_spin(unsigned int dots)
{
	this->dots = dots;
	this->larmor_frequency = new realnum[dots]();
	this->rabi_frequency = new realnum[dots]();
	this->charging_energy = new realnum[dots]();
	this->singlet_triplet_energy = new realnum[dots]();
	this->microwave_control = new realnum[dots]();
	this->detuning_control = new realnum[dots]();
	this->tunnel_control = new realnum[dots*dots]();
}

spine::systems::system_spin::~system_spin()
{
	delete[] larmor_frequency;
	delete[] rabi_frequency;
	delete[] charging_energy;
	delete[] singlet_triplet_energy;
	delete[] microwave_control;
	delete[] detuning_control;
	delete[] tunnel_control;
#ifdef PLOT
	if (p != nullptr) {
		for (unsigned int dot = 0; dot < dots; dot++)
			delete p[dot];
		delete[] p;
	}
#endif
}

int spine::systems::system_spin::initializeIndex()
{
	unsigned int * states = new unsigned int[dots];
	for (unsigned int dot = 0; dot < dots; dot++)
		states[dot] = STATE_0;
	int index = getIndex(states);
	delete[] states;
	return index;
}

void spine::systems::system_spin::initialize(complex * state)
{
	// Zero the state vector
	initZero(getDimension(), state);

	// Set the state at the index of the ground state to 1
	int index = initializeIndex();
	if (index >= 0)
		state[index].real = (realnum) 1.0;
}

void spine::systems::system_spin::measure(complex * state, realnum t, realnum * Px, realnum * Py, realnum * Pz)
{
	// For every dot
	for (unsigned int dot = 0; dot < dots; dot++) {
		Px[dot] = 0;
		Py[dot] = 0;
		Pz[dot] = 0;

		// Measure in rotating frame at time t if requested
		complex rot_x = { 1, 0 };
		complex rot_y = { 0, 1 };
		if (t != 0) {
#ifdef DOUBLE_PRECISION
			rot_x.real = cos(larmor_frequency[dot] * t);
			rot_x.imag = sin(larmor_frequency[dot] * t);
			rot_y.real = cos(larmor_frequency[dot] * t + M_PI / 2.0);
			rot_y.imag = sin(larmor_frequency[dot] * t + M_PI / 2.0);
#else
			rot_x.real = cosf(larmor_frequency[dot] * t);
			rot_x.imag = sinf(larmor_frequency[dot] * t);
			rot_y.real = cosf(larmor_frequency[dot] * t + (realnum)M_PI / 2.0f);
			rot_y.imag = sinf(larmor_frequency[dot] * t + (realnum)M_PI / 2.0f);
#endif
		}

		// Sum over all indices where this dot in the requested state
		vector<unsigned int> * indices0 = getIndexMeasurement(dot, STATE_0);
		vector<unsigned int> * indices1 = getIndexMeasurement(dot, STATE_1);
		vector<unsigned int>::iterator it1 = indices1->begin();
		for (vector<unsigned int>::iterator it0 = indices0->begin(); it0 != indices0->end(); ++it0) {
			unsigned int index0 = *it0;
			unsigned int index1 = *it1;
			complex tmp;

			// |(state[index0] + rot_x * state[index1]) / sqrt(2)| ^ 2
			tmp.real = state[index0].real + rot_x.real * state[index1].real - rot_x.imag * state[index1].imag;
			tmp.imag = state[index0].imag + rot_x.imag * state[index1].real + rot_x.real * state[index1].imag;
			Px[dot] += abs2(tmp) / ((realnum) 2.0);

			// |(state[index0] + rot_y * state[index1]) / sqrt(2)| ^ 2
			tmp.real = state[index0].real + rot_y.real * state[index1].real - rot_y.imag * state[index1].imag;
			tmp.imag = state[index0].imag + rot_y.imag * state[index1].real + rot_y.real * state[index1].imag;
			Py[dot] += abs2(tmp) / ((realnum) 2.0);

			// |state[index0]| ^ 2
			Pz[dot] += abs2(state[index0]);

			++it1;
		}
	}
}

void spine::systems::system_spin::measure(complex * state, realnum * Px, realnum * Py, realnum * Pz)
{
	return measure(state, 0, Px, Py, Pz);
}

void spine::systems::system_spin::measureST(complex * state, realnum * Pn, realnum * Ps, realnum * Pt)
{
	// For every dot
	for (unsigned int dot = 0; dot < dots; dot++) {
		Pn[dot] = 0;
		Ps[dot] = 0;
		Pt[dot] = 0;

		// Sum over all indices where this dot in the requested state
		vector<unsigned int> * indices = getIndexMeasurement(dot, STATE_N);
		for (vector<unsigned int>::iterator it = indices->begin(); it != indices->end(); ++it)
			Pn[dot] += abs2(state[*it]);
		indices = getIndexMeasurement(dot, STATE_S);
		for (vector<unsigned int>::iterator it = indices->begin(); it != indices->end(); ++it)
			Ps[dot] += abs2(state[*it]);
		indices = getIndexMeasurement(dot, STATE_T0);
		for (vector<unsigned int>::iterator it = indices->begin(); it != indices->end(); ++it)
			Pt[dot] += abs2(state[*it]);
		indices = getIndexMeasurement(dot, STATE_TP);
		for (vector<unsigned int>::iterator it = indices->begin(); it != indices->end(); ++it)
			Pt[dot] += abs2(state[*it]);
		indices = getIndexMeasurement(dot, STATE_TM);
		for (vector<unsigned int>::iterator it = indices->begin(); it != indices->end(); ++it)
			Pt[dot] += abs2(state[*it]);
	}
}

#ifdef PLOT

void spine::systems::system_spin::plotSetup(unsigned int points, double xmin, double xmax)
{
	this->points = points;
	this->xmin = xmin;
	this->xmax = xmax;
}

void spine::systems::system_spin::plotAddU(complex * U, realnum t, bool plot_lab)
{
	unsigned int dim = getDimension();
	complex * statetmp = allocComplex(dim);
	complex * state = allocComplex(dim);
	
	// Update the overall operation
	initialize(state);
	initCopy(dim, statetmp, state);
	multiplyVector(dim, U, statetmp, state);
	plotAdd(state, t, plot_lab);

	freeComplex(state);
	freeComplex(statetmp);
}

void spine::systems::system_spin::plotAdd(complex * state, realnum t, bool plot_lab)
{
	realnum * Px = nullptr;
	realnum * Py = nullptr;
	realnum * Pz = nullptr;
	if (plot_lab)
	{
		Px = new realnum[dots];
		Py = new realnum[dots];
		Pz = new realnum[dots];
		measure(state, Px, Py, Pz);
	}
	realnum * Pxr = new realnum[dots];
	realnum * Pyr = new realnum[dots];
	realnum * Pzr = new realnum[dots];
	measure(state, t, Pxr, Pyr, Pzr);

	if (p == nullptr)
	{
		p = new Plot*[dots];
		for (unsigned int dot = 0; dot < dots; dot++)
		{
			wchar_t title[10];
			swprintf(title, 10, L"Qubit %d", dot);
			p[dot] = new Plot(title, points);
			if (plot_lab)
			{
				p[dot]->add(255, 192, 192, xmin, xmax, 0, 1);
				p[dot]->add(192, 255, 192, xmin, xmax, 0, 1);
			}
			p[dot]->add(255, 0, 0, xmin, xmax, 0, 1);
			p[dot]->add(0, 255, 0, xmin, xmax, 0, 1);
			p[dot]->add(0, 0, 255, xmin, xmax, 0, 1);
		}
	}
	for (unsigned int dot = 0; dot < dots; dot++)
	{
		if (plot_lab)
		{
			p[dot]->get(0)->add(t, Px[dot]);
			p[dot]->get(1)->add(t, Py[dot]);
			p[dot]->get(2)->add(t, Pxr[dot]);
			p[dot]->get(3)->add(t, Pyr[dot]);
			p[dot]->get(4)->add(t, Pzr[dot]);
		}
		else
		{
			p[dot]->get(0)->add(t, Pxr[dot]);
			p[dot]->get(1)->add(t, Pyr[dot]);
			p[dot]->get(2)->add(t, Pzr[dot]);
		}
	}

	if (plot_lab)
	{
		delete[] Px;
		delete[] Py;
		delete[] Pz;
	}
	delete[] Pxr;
	delete[] Pyr;
	delete[] Pzr;
}

void spine::systems::system_spin::plot()
{
	if (p != nullptr)
		for (unsigned int dot = 0; dot < dots; dot++)
			p[dot]->redraw();
}

#endif

void spine::systems::system_spin::setLarmorFrequency(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		larmor_frequency[dot] = value;
}

void spine::systems::system_spin::setLarmorFrequency(unsigned int dot, realnum value)
{
	larmor_frequency[dot] = value;
}

realnum spine::systems::system_spin::getLarmorFrequency()
{
	return larmor_frequency[0];
}

realnum spine::systems::system_spin::getLarmorFrequency(unsigned int dot)
{
	return larmor_frequency[dot];
}

void spine::systems::system_spin::setRabiFrequency(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		rabi_frequency[dot] = value;
}

void spine::systems::system_spin::setRabiFrequency(unsigned int dot, realnum value)
{
	rabi_frequency[dot] = value;
}

realnum spine::systems::system_spin::getRabiFrequency()
{
	return rabi_frequency[0];
}

realnum spine::systems::system_spin::getRabiFrequency(unsigned int dot)
{
	return rabi_frequency[dot];
}

void spine::systems::system_spin::setChargingEnergy(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		charging_energy[dot] = value;
}

void spine::systems::system_spin::setChargingEnergy(unsigned int dot, realnum value)
{
	charging_energy[dot] = value;
}

realnum spine::systems::system_spin::getChargingEnergy()
{
	return charging_energy[0];
}

realnum spine::systems::system_spin::getChargingEnergy(unsigned int dot)
{
	return charging_energy[dot];
}

void spine::systems::system_spin::setSingletTripletEnergy(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		singlet_triplet_energy[dot] = value;
}

void spine::systems::system_spin::setSingletTripletEnergy(unsigned int dot, realnum value)
{
	singlet_triplet_energy[dot] = value;
}

realnum spine::systems::system_spin::getSingletTripletEnergy()
{
	return singlet_triplet_energy[0];
}

realnum spine::systems::system_spin::getSingletTripletEnergy(unsigned int dot)
{
	return singlet_triplet_energy[dot];
}

void spine::systems::system_spin::setMicrowaveControl(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		microwave_control[dot] = value;
}

void spine::systems::system_spin::setMicrowaveControl(unsigned int dot, realnum value)
{
	microwave_control[dot] = value;
}

void spine::systems::system_spin::setDetuningControl(realnum value)
{
	for (unsigned int dot = 0; dot < dots; dot++)
		detuning_control[dot] = value;
}

void spine::systems::system_spin::setDetuningControl(unsigned int dot, realnum value)
{
	detuning_control[dot] = value;
}

void spine::systems::system_spin::setTunnelControl(realnum value)
{
	for (unsigned int dota = 0; dota < dots; dota++) {
		for (unsigned int dotb = 0; dotb < dots; dotb++) {
			tunnel_control[dota + dots * dotb] = value;
			tunnel_control[dotb + dots * dota] = value;
		}
	}
}

void spine::systems::system_spin::setTunnelControl(unsigned int dota, unsigned int dotb, realnum value)
{
	tunnel_control[dota + dots * dotb] = value;
	tunnel_control[dotb + dots * dota] = value;
}
